### PR TITLE
Remove 'duplicate' fixed chains.

### PR DESCRIPTION
**Pull request by taknira**
...where duplicate means that one chain is a subchain of another.
e.g. A -> B is a subchain of A -> B -> C.  However, we do not
consider A -> B to be a subchain of Z -> A -> B.  The subchains
must at the leaf.

This is to prevent the preloader logging more chain than it needs to.

### CODE REVIEW

File: go/fixchain/fixer.go [hunk 1]
```diff
@@ -76,6 +79,48 @@ func (f *Fixer) updateCounters(ferrs []*FixError) {
...
+type chainSlice struct {
+	chains [][]*x509.Certificate
+}
+
+// sort.Sort(data Interface) for chainSlice - uses data.Len, data.Less & data.Swap.
+func (c chainSlice) Len() int { return len(c.chains) }
+func (c chainSlice) Less(i, j int) bool {
+	chi := c.chains[i]
+	chj := c.chains[j]
+	for k := 0; k < int(math.Min(float64(len(chi)), float64(len(chj)))); k++ {
```
Comment (benlaurie) [inline go/fixchain/fixer.go]: Ugh. This is an abuse of math.Min().

File: go/fixchain/fixer.go [hunk 2]
```diff
@@ -76,6 +79,48 @@ func (f *Fixer) updateCounters(ferrs []*FixError) {
 	atomic.AddUint32(&f.reconstructed, 1)
 }
 
+type chainSlice struct {
+	chains [][]*x509.Certificate
+}
+
+// sort.Sort(data Interface) for chainSlice - uses data.Len, data.Less & data.Swap.
```
Comment (benlaurie) [inline go/fixchain/fixer.go]: Explain what order this sorts into.

File: go/fixchain/fixer.go [hunk 3]
```diff
@@ -76,6 +79,48 @@ func (f *Fixer) updateCounters(ferrs []*FixError) {
...
+	}
+	return len(chi) < len(chj)
+}
+func (c chainSlice) Swap(i, j int) {
+	t := c.chains[i]
+	c.chains[i] = c.chains[j]
+	c.chains[j] = t
+}
+
+func removeSuperChains(chains [][]*x509.Certificate) [][]*x509.Certificate {
```
Comment (benlaurie) [inline go/fixchain/fixer.go]: The algorithm this uses is slightly subtle. Explain it.

File: go/fixchain/fixer.go [hunk 4]
```diff
@@ -105,9 +157,9 @@ func (f *Fixer) logStats() {
 	go func() {
 		for _ = range t.C {
 			log.Printf("fixers: %d active, %d reconstructed, "+
-				"%d not reconstructed, %d fixed, %d not fixed",
+				"%d not reconstructed, %d fixed, %d not fixed, %d valid chains sent on chan",
```
Comment (benlaurie) [inline go/fixchain/fixer.go]: remove "on chan".

File: go/fixchain/fixer.go (hunk 1 above)
Comment (taknira) [inline go/fixchain/fixer.go]: Abuse removed

File: go/fixchain/fixer.go (hunk 2 above)
Comment (taknira) [inline go/fixchain/fixer.go]: Done.

File: go/fixchain/fixer.go (hunk 3 above)
Comment (taknira) [inline go/fixchain/fixer.go]: Done.

File: go/fixchain/fixer.go (hunk 4 above)
Comment (taknira) [inline go/fixchain/fixer.go]: Done.

File: go/fixchain/fixer_test.go [hunk 5]
```diff
@@ -126,7 +208,7 @@ func TestUpdateCounters(t *testing.T) {
 		for _, err := range test.errors {
 			ferrs = append(ferrs, &FixError{Type: err})
 		}
-		f.updateCounters(ferrs)
+		f.updateCounters(nil, ferrs)
```
Comment (benlaurie) [inline go/fixchain/fixer_test.go]: This appears to be an unrelated change.

Comment (taknira) [inline go/fixchain/fixer_test.go]: Yes, it should have gone in the last PR (1139), but that has already been merged.  It was my mistake.
